src/sync.ts

```diff
@@ -1909,54 +1922,59 @@ export function _createAndReEmitRoom(client: MatrixClient, roomId: string, opts:
  *
- * Converts the events into `MatrixEvent`s, and emits appropriate {@link ClientEvent.ToDeviceEvent} events.
+ * Converts the to-device message and OlmEncryptionIngo and emits appropriate {@link ClientEvent.ReceivedToDeviceMessage} events.
+ * Converts the events into `MatrixEvent`s, and emits the now deprecated {@link ClientEvent.ToDeviceEvent} events for compatibility.
  * */
-export function processToDeviceMessages(toDeviceMessages: IToDeviceEvent[], client: MatrixClient): void {
+export function processToDeviceMessages(toDeviceMessages: ReceivedToDeviceMessage[], client: MatrixClient): void {
     const cancelledKeyVerificationTxns: string[] = [];
     toDeviceMessages
-        .map(mapToDeviceEvent)
-        .map((toDeviceEvent) => {
+        .map((processedMessage) => {
             // map is a cheap inline forEach
             // We want to flag m.key.verification.start events as cancelled
             // if there's an accompanying m.key.verification.cancel event, so
             // we pull out the transaction IDs from the cancellation events
             // so we can flag the verification events as cancelled in the loop
             // below.
-            if (toDeviceEvent.getType() === "m.key.verification.cancel") {
-                const txnId: string = toDeviceEvent.getContent()["transaction_id"];
+            if (processedMessage.message.type === "m.key.verification.cancel") {
+                const txnId: string = processedMessage.message.content["transaction_id"];
                 if (txnId) {
                     cancelledKeyVerificationTxns.push(txnId);
                 }
             }
 
             // as mentioned above, .map is a cheap inline forEach, so return
             // the unmodified event.
-            return toDeviceEvent;
+            return processedMessage;
         })
-        .forEach(function (toDeviceEvent) {
-            const content = toDeviceEvent.getContent();
-            if (toDeviceEvent.getType() == "m.room.message" && content.msgtype == "m.bad.encrypted") {
-                // the mapper already logged a warning.
-                logger.log("Ignoring undecryptable to-device event from " + toDeviceEvent.getSender());
-                return;
-            }
-
-            if (
-                toDeviceEvent.getType() === "m.key.verification.start" ||
-                toDeviceEvent.getType() === "m.key.verification.request"
-            ) {
-                const txnId = content["transaction_id"];
-                if (cancelledKeyVerificationTxns.includes(txnId)) {
-                    toDeviceEvent.flagCancelled();
+        .forEach(function (processedEvent) {
+            // For backwards compatibility, we emit the event as a MatrixEvent
+            {
+                const toDeviceEvent = processedEvent.message;
+                const content = toDeviceEvent.content;
+                const deprecatedCompatibilityEvent = new MatrixEvent(Object.assign({}, toDeviceEvent));
```

Comment:
what is the purpose of the `Object.assign` here? might be worth adding a comment?

Reply:
added. 
copy here:
```
// The message is cloned before being passed to the MatrixEvent constructor, because
// the `makeEncrypted` method will mutate the type and content properties of the original message and will interfere
// with the emitted event for `ReceivedToDeviceMessage`.